### PR TITLE
Metrics xmpp stanzas count

### PR DESCRIPTION
**Pull request by janciesla8818**
This PR extends functionality of collecting metrics. It allows to gather the number of sent and received:
* Message Stanzas
* Iq Stanzas
* Presence Stanzas

The information is retrieved from exometer. The metrics are sending information regarding number of stanzas since the last report and total amount of stanzas as reported by exometer.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[7999.1](https://travis-ci.org/esl/MongooseIM/jobs/660231567) / Erlang 22.0 / small_tests / 4294622c4b662be75172a87383f37dcd71dccfcd
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2672/7999/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2672/7999/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-03-09_16.26.24/lib.mongooseim.logs/run.2020-03-09_16.26.24/suite.log.html)

File: src/system_metrics/mongoose_system_metrics_collector.erl [hunk 1]
```diff
@@ -188,3 +189,35 @@ get_outgoing_pools() ->
     [#{report_name => outgoing_pools,
        key => type,
        value => Type} || {Type, _, _, _, _} <- OutgoingPools].
+
+get_xmpp_stanzas_count() ->
+    Hosts = ejabberd_config:get_global_option(hosts),
```
Comment (michalwski) [inline src/system_metrics/mongoose_system_metrics_collector.erl]: It's possible to configure all metrics to be `global`. In this case looking at `hosts` will return nothing. Take a look at the `all_metrics_are_global` configuration option.

File: src/system_metrics/mongoose_system_metrics_collector.erl [hunk 2]
```diff
@@ -188,3 +189,35 @@ get_outgoing_pools() ->
...
+    StanzaTypes = [xmppMessageSent, xmppMessageReceived, xmppIqSent,
+                   xmppIqReceived, xmppPresenceSent, xmppPresenceReceived],
+    NewCount = [count_stanzas(Hosts, StanzaType) || StanzaType <- StanzaTypes],
+    PrevReport = mongoose_system_metrics_file:read(),
```
Comment (michalwski) [inline src/system_metrics/mongoose_system_metrics_collector.erl]: How hard would it be to store the prev value in the state of the main process triggering the metrics collection? I'm not sure if read from file is the best solution, the file can be removed or changed at any time.

File: src/system_metrics/mongoose_system_metrics_collector.erl [hunk 3]
```diff
@@ -188,3 +189,35 @@ get_outgoing_pools() ->
...
+    StanzasCount = calculate_stanza_rate(PrevReport, NewCount),
+    [#{report_name => StanzaType,
+       key => Total,
+       value => Increment} || {StanzaType, Total, Increment} <- StanzasCount].
+
+count_stanzas(Hosts, StanzaType) ->
+    FullExometerResults = [exometer:get_value([Host, StanzaType]) || Host <- Hosts],
```
Comment (michalwski) [inline src/system_metrics/mongoose_system_metrics_collector.erl]: As far as I know even if MongooseIM is configured to have all metrics as global, they have the same level. It's for instance `[Host, xmppMessagesSent]` for per host metrics or `[global, xmppMessagesSent]` if MongooseIM is configured to have all metrics as global. We could use the following exometer API:
```erlang
exometer:get_values(['_', xmppMessageSent])
```
to get all the values despite number of hosts or if metrics are global or not.

[the remark this answers is not in the document]
Comment (NelsonVides) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: Or, yeah, that's also a good way to do so https://github.com/esl/MongooseIM/pull/2672#discussion_r390201794 :)

[the remark this answers is not in the document]
Comment (michalwski) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: I don't think testing the edge case is necessary.

File: big_tests/tests/service_mongoose_system_metrics_SUITE.erl [hunk 4]
```diff
@@ -291,6 +293,9 @@ transport_mechanisms_are_reported(_Config) ->
 outgoing_pools_are_reported(_Config) ->
     mongoose_helper:wait_until(fun outgoing_pools_are_reported/0, true).
 
+xmpp_stanzas_counts_are_reported(_Config) ->
+    mongoose_helper:wait_until(fun xmpp_stanzas_counts_are_reported/0, true).
```
Comment (michalwski) [inline big_tests/tests/service_mongoose_system_metrics_SUITE.erl]: I think we could extend the test by connecting some users, sending some stanzas and checking if the reported value changed. Is is possible? Currently the test only check if the metric is reported, but it can be reported with wrong values. It'd be good to verify that.

File: src/system_metrics/mongoose_system_metrics_collector.erl (hunk 1 above)
Comment (janciesla8818) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: I'm thinking getting a list of local and global values from mongoose_metrcis:get_value , filter out the errors and proceed with extracting values.

File: src/system_metrics/mongoose_system_metrics_collector.erl (hunk 2 above)
Comment (janciesla8818) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: Returning just these values wouldn't be that hard, just could be a bit messy. What do you think of storing full last report in the state of the main process? Collecting metrics would take Previous report as an argument.

File: src/system_metrics/mongoose_system_metrics_collector.erl (hunk 3 above)
Comment (janciesla8818) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: yep, this would be the best way to get this data

File: src/system_metrics/mongoose_system_metrics_collector.erl (hunk 1 above)
Comment (michalwski) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: As I mentioned in my other comment, the following trick from exometer API could be used: `exometer:get_values(['_', xmppMessageSent])`. In this case, you don't have to bother what is the configuration, what are the hosts or if there were errors reading a metric.

Comment (janciesla8818) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: Sure, sorry for the confusion. I was trying to address the comments one by one reading from the top and only came later to the one you mention. But totally agree, will get start working on this approach.

File: src/system_metrics/mongoose_system_metrics_collector.erl (hunk 2 above)
Comment (michalwski) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: I think we could store in the state only these values which are needed for the next report. I'm not sure there is a need to store full report. On the other hand storing full report may indeed be cleaner.

Comment (janciesla8818) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: Storing full report would make it easier to understand on what is stored there. It could either be 
* full report
* short report with values that might be needed for the next one
* explicit tuples {metric, count} and having each value stored explicitly in the state
which do you think would make most sense to implement?

Comment (michalwski) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: Let's go for the simplicity and store the full report in the state.

**Comment by mongoose-im** [comment]
[8004.1](https://travis-ci.org/esl/MongooseIM/jobs/660753372) / Erlang 22.0 / small_tests / 245535500f898ed84a04ad06cf2ef590931dc141
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2672/8004/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2672/8004/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-03-10_18.37.29/lib.mongooseim.logs/run.2020-03-10_18.37.30/suite.log.html)

**Comment by mongoose-im** [comment]
[8005.1](https://travis-ci.org/esl/MongooseIM/jobs/660789264) / Erlang 22.0 / small_tests / fd09bcb20b34b112a1e0db640c94947de874abb3
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2672/8005/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2672/8005/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-03-10_20.05.51/lib.mongooseim.logs/run.2020-03-10_20.05.51/suite.log.html)

File: src/system_metrics/mongoose_system_metrics_file.erl [hunk 5]
```diff
@@ -15,3 +15,4 @@ save(Reports) ->
     JSON = jiffy:encode(Reports, [pretty]),
     file:write_file(location(), JSON),
     ok.
+
```
Comment (NelsonVides) [inline src/system_metrics/mongoose_system_metrics_file.erl]: We probably don't need this new line :roll_eyes:

File: src/system_metrics/service_mongoose_system_metrics.erl [hunk 6]
```diff
@@ -25,7 +25,8 @@
 
 -export([verify_if_configured/0]).
 
--record(system_metrics_state, {report_after, reporter_monitor = none, reporter_pid = none}).
+-record(system_metrics_state, {report_after, reporter_monitor = none,
+                               reporter_pid = none, prev_report = []}).
```
Comment (NelsonVides) [inline src/system_metrics/service_mongoose_system_metrics.erl]: I'd give this some type information, like
```erlang
-record(system_metrics_state,
        {
         report_after :: non_neg_integer(),
         reporter_monitor = none :: reference(),
         reporter_pid = none :: pid(),
         prev_report = []
        }).
```
And it itches me to leave `prev_report` untyped but ok, no need to export types from other modules and so on, this is sent as a message anyway so nor dialyzer nor gradualizer will do anything anyway :man_shrugging:

File: src/system_metrics/service_mongoose_system_metrics.erl [hunk 7]
```diff
@@ -66,14 +67,17 @@ init(Args) ->
...
     case get_client_id() of
         {ok, ClientId} ->
             {Pid, Monitor} = spawn_monitor(
                 fun() ->
-                    Reports = mongoose_system_metrics_collector:collect(),
+                    Reports = mongoose_system_metrics_collector:collect(PrevReport),
                     mongoose_system_metrics_sender:send(ClientId, Reports),
-                    mongoose_system_metrics_file:save(Reports)
+                    mongoose_system_metrics_file:save(Reports),
+                    ServicePid ! {prev_report, Reports}
```
Comment (NelsonVides) [inline src/system_metrics/service_mongoose_system_metrics.erl]: Actually, while this is perhaps more readable, it also means this new process sends two messages to the master process, one with this bang, and another one as the `'DOWN'` message when it dies. If here we do instead something like
```erlang
    exit({report, Report})
```
Only one message will be sent, with the Report, which can be caught in the `handle_info` as it is done already, with something like
```erlang
 handle_info({'DOWN', CollectorMonitor, _, _, {report, Report}}, 
```
Just saying, maybe is an option? (@michalwski)

File: big_tests/tests/service_mongoose_system_metrics_SUITE.erl [hunk 8]
```diff
@@ -291,6 +303,16 @@ transport_mechanisms_are_reported(_Config) ->
 outgoing_pools_are_reported(_Config) ->
     mongoose_helper:wait_until(fun outgoing_pools_are_reported/0, true).
 
+xmpp_stanzas_counts_are_reported(Config) ->
+    escalus:story(Config, [{alice,1}, {bob,1}], fun(Alice, Bob) ->
+        escalus:send(Alice, escalus_stanza:chat_to(Bob, <<"COVID-19">>)),
+        escalus:assert(is_chat_message, [<<"COVID-19">>],
+                       escalus:wait_for_stanza(Bob))
+
```
Comment (NelsonVides) [inline big_tests/tests/service_mongoose_system_metrics_SUITE.erl]: not needed empty line :roll_eyes:

File: big_tests/tests/service_mongoose_system_metrics_SUITE.erl [hunk 9]
```diff
@@ -291,6 +303,16 @@ transport_mechanisms_are_reported(_Config) ->
 outgoing_pools_are_reported(_Config) ->
     mongoose_helper:wait_until(fun outgoing_pools_are_reported/0, true).
 
+xmpp_stanzas_counts_are_reported(Config) ->
+    escalus:story(Config, [{alice,1}, {bob,1}], fun(Alice, Bob) ->
+        escalus:send(Alice, escalus_stanza:chat_to(Bob, <<"COVID-19">>)),
+        escalus:assert(is_chat_message, [<<"COVID-19">>],
```
Comment (NelsonVides) [inline big_tests/tests/service_mongoose_system_metrics_SUITE.erl]: WTF, really? :joy: I wonder who'll wonder about this? Ourselves in some time when we have to wikipedia this thing because we forgot what was it. Or our survivors who will find your sense of humour politically incorrect. I'm signing for this :joy:

**Comment by NelsonVides** [comment]
Oh damn, @michalwski you merged the PR just while I was finishing my review xD

File: src/system_metrics/service_mongoose_system_metrics.erl (hunk 7 above)
Comment (michalwski) [reply src/system_metrics/service_mongoose_system_metrics.erl]: I see what you mean @NelsonVides. At this moment I'm not convinced sending a message with result via `exit` is clear. In my opinion it may lead to confusion.
